File: implementations/react/src/patterns/structural/facade/subsystems/crm-subsystem.ts
import type { Customer, Slot } from "../types/types";

/**
 * Subsistema de CRM que maneja clientes y actividades.
 */

export class CRMService {
  async upsertCustomer(c: Customer) {
    return { customerId: "cus_" + Math.random().toString(36).slice(2), ...c };
  }
  async createActivity(opts: { customerId: string; slot: Slot; txId?: string }) {
    return { activityId: "act_" + Math.random().toString(36).slice(2), ...opts };
  }
}

File: implementations/flutter/design_patterns_flutter/lib/patterns/structural/facade/subsystems/crm_subsystem.dart
import '../types.dart';

class CrmCustomer {
  final String customerId;
  final String email;
  final String? phone;

  CrmCustomer({required this.customerId, required this.email, this.phone});
}

class CrmActivity {
  final String activityId;
  CrmActivity(this.activityId);
}

class CRMService {
  int _id = 1;

  Future<CrmCustomer> upsertCustomer(Customer c) async {
    // Simulate network delay
    await Future.delayed(const Duration(milliseconds: 120));
    return CrmCustomer(customerId: 'cust-${_id++}', email: c.email, phone: c.phone);
  }

  Future<CrmActivity> createActivity(Map<String, dynamic> opts) async {
    await Future.delayed(const Duration(milliseconds: 80));
    return CrmActivity('act-${DateTime.now().millisecondsSinceEpoch}');
  }
}
